docs/source/community/rfc/rfc-3.rst

```diff
@@ -0,0 +1,112 @@
+update of version or standard requirements is described. The reasoning behind
+a rolling update scheme is that it has become increasingly evident that C89
+is becoming outdated and creating a less than optimal development environment
+for contributors. It has been noted that the most commonly used compilers all
+now support more recent versions of C, so the strict usage of C89 is no longer
+as critical as it used to be.
+
+Similarly, rolling updates to other tools and libraries that PROJ depend on
+will ensure that the code base can be kept modern and in line with the rest of
+the open source software ecosphere.
+
+C and C++
+-------------------------------------------------------------------------------
+
+Following :ref:`RFC2 <rfc2>` PROJ is written in both C and C++. At the time of writing
+the core library is C based and the code described in RFC2 is written in C++.
+For the foreseeable future this will be upheld but it is likely that a shift
+towards C++ will happen some time in the not-so-distant future.
+
+Both the C and C++ standards are updated with regular intervals. After an
+update of a standard it takes time for compiler manufacturers to implement the
+standards fully, which makes adaption of new standards potentially troublesome
+if done too soon. On the other hand, waiting too long to adopt new standards
+will eventually make the code base feel old and new contributors are more likely
+to stay away because they don't want to work using tools of the past. With a
+rolling update scheme both concerns can be managed by always staying behind the
+most recent standard, but not so far away that potential contributors are scared
+away. Keeping a policy of always lagging behind be two iterations of the standard
+is thought to be the best comprise between the two concerns.
+
+C comes in four ISO standardised varieties: C89, C99, C11, C18. In this document
+we refer to their informal names for ease of reading. C++ exists in five
+varieties: C++98, C++03, C++11, C++14, C++17. Before adoption of this RFC PROJ
+uses C89 and C++11. For C that means that the used standard is three iterations
+behind the most recent standard. C++ is two iterations behind. After adoption of
+this RFC the required standard for compilation of PROJ will always be two
+iterations behind the most recent standard. This will result in a change of C
+standard immediately after adoption of the RFC. When a new standard for either C
+or C++ is released PROJ should change its requirement to the next standard in
+the line. For C++ that means a change in standard roughly every three years, for
+C the periods between standard updates is expected to be longer. Adaptation of
+new programming language standards should be coordinated with a major version
+release of PROJ. The initial switch to C99 will be coordinated with the release
+of PROJ 6.0.0. Adaptation of C++14 is possible with PROJ 7.0.0, provided that
```

Comment:
Beyond the age of the standards, I think we should also look at what the compilers actually implement (and Microsoft is the one traditionally lagging behind). It might be also prudent to have a sort of X-version-behind rule for the compilers themselves (e.g with C++11, that is a 1-version-behind for Visual Studio, since you need VS >= 2015) . And perhaps take into consideration what is available in LTS versions of Linux distributions and in the CI environments we use

Reply:
> Beyond the age of the standards, I think we should also look at what the compilers actually implement (and Microsoft is the one traditionally lagging behind). It might be also prudent to have a sort of X-version-behind rule for the compilers themselves (e.g with C++11, that is a 1-version-behind for Visual Studio, since you need VS >= 2015) . 

Good point. My thinking was that by lagging behind two iterations of the standard compilers would be more or less caught up. It seems to me that Microsoft has started moving a bit faster in that regard, but maybe that is just my imagination? Both GCC and LLVM seem to be quite good at following along the standards.
Which compilers should be considered? MSVC, GCC, LLVM?

Reply:
FYI, VS2017 conforms to C++11, 14 and 17 now with -std=c++14 being the default mode.

Reply:
I think the discussion about compiler versions should be oriented toward the C++ standard it supports and nothing else. Otherwise, the support matrix becomes convoluted and every developer ends up being responsible for cross-checking their development against the entire thing.

The matrix as defined right now covers "PROJ requires C++11" as a statement. I think there should be an RFC vote to change that to "PROJ requires C++14" or any other future revision.

Reply:
Thanks for your input guys. So summing up, it sounds like there's some agreement that we shouldn't just automatically switch to a newer revision of a standard because a new one is released. At least not without confirming that there's broad support of said standard revision first. That was sort also what I was going for but I think I'll have to reword the text to make that clearer. The example about switching to C++14 with version 7 was just that, an example demonstrating the idea. I would not suggest a switch at that point, but it might be beneficial for PROJ 8. I Agree that the PSC should confirm an update to this RFC before moving to a new C/C++ standard.